### PR TITLE
cylc domain

### LINKED ISSUE #11: cylc domain
Add a sphinx domain for cylc configurations.

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #11 
Partly addresses #12 

It's hard-to-review code-bomb time! Sorry! I'll try to comment this better to help.

The code is documented in module docstrings, you can build it with:

```console
$ make clean html
$ firefox _build/html/index.html
```

(note: the docs use a strict build so building the docs also tests them)

### Cylc Domain

Create a [Sphinx domain](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html) for documenting Cylc configurations. This allows us to document settings as objects which we can reference throughout the Cylc documentation (and via inter-sphinx any Sphinx project).

#### Benefits:

* A formal way to document configurations.
* A formal way to reference configurations internally and externally.
* References get checked on build so if we reference a configuration that has been removed or renamed we will get an error.

#### Definition examples:

Cylc configurations can be documented manually like this:

```rst
.. cylc:conf:: myconf

   Description of myconf

   .. cylc:section:: foo

      Description of foo

      .. cylc:setting:: bar

        :type: string
        :default: pub

        Description of bar.
```

#### Reference examples:

Once defined objects can be referenced like this:

* `:cylc:conf:'suite.rc'`
* `:cylc:section:'[runtime]'`
* `:cylc:setting:'[scheduling]max active cycle points'`

### Cylc AutoDocumenter

This PR also adds a POC autodocumenter which can document Cylc configurations from `SPEC` dictionaries (at present only type, default and options are available). This will need to be re-worked when we build documentation into Cylc configurations.

#### Parsec examples:

Parsec types can be be autodocumented like so (requires https://github.com/cylc/cylc-flow/pull/3439):

```rst
.. auto-cylc-type::
   cylc.flow.parsec.validate.ParsecValidator.V_TYPE_HELP
   cylc.flow.parsec.validate.CylcConfigValidator.V_TYPE_HELP
```

#### Cylc examples:

Cylc configs can be autodocumented like so:

```rst
.. auto-cylc-conf:: suite cylc.flow.cfgspec.suite.SPEC
```

### CODE REVIEW

**Comment by oliver-sanders** [comment]
The Travis build is failing due to an interaction between hieroglyph + docutils==0.1.6

**Comment by oliver-sanders** [comment]
NOTE:

This can go in before https://github.com/cylc/cylc-flow/issues/3253 is attempted, the autodocumenter will need to be updated to work with whatever replaces the `SPEC` dictionary.

File: cylc/sphinx_ext/diff_selection/__init__.py [hunk 1]
```diff
@@ -5,8 +5,11 @@
 * Makes ``-`` lines non-selectable.
 * Makes ``+`` symbol non-selectable.
 
-Extension is automatically loaded for all code diffs when it is added to
-a project's Sphinx extensions. No directives required.
+.. note::
+
+   Extension is automatically loaded when added to extensions, no directives or
+   confiruations required.
```
Comment (kinow) [inline cylc/sphinx_ext/diff_selection/__init__.py]: s/confiruations /configurations

File: cylc/sphinx_ext/grid_table/__init__.py [hunk 2]
```diff
@@ -1,42 +1,29 @@
...
-Re-styles the Sphinx built-in :rst:dir:`list-table` directive.
+Re-styles the Sphinx built-in ``list-table`` directive.
 
-Extension is automatically loaded for all code diffs when it is added to
-a project's Sphinx extensions. No directives required.
+.. note::
 
-.. rubric:: Example
+   Extension is automatically loaded when added to extensions, no directives or
+   confiruations required.
```
Comment (kinow) [inline cylc/sphinx_ext/grid_table/__init__.py]: s/confiruations/configurations

File: cylc/sphinx_ext/hieroglyph_addons/__init__.py [hunk 3]
```diff
@@ -8,6 +8,13 @@
 
 .. nextslide::
 
+.. note::
+
+   extension is automatically loaded when added to extensions, no directives or
+   confiruations required.
```
Comment (kinow) [inline cylc/sphinx_ext/hieroglyph_addons/__init__.py]: s/confiruations/configurations

File: cylc/sphinx_ext/hieroglyph_patch/__init__.py [hunk 4]
```diff
@@ -29,6 +29,12 @@
 to patch extensions automatically but in the interests of keeping the hack to a
 minimum this hard-coded extension should suffice.
 
+.. note::
+
+   extension is automatically loaded when added to extensions, no directives or
+   confiruations required.
```
Comment (kinow) [inline cylc/sphinx_ext/hieroglyph_patch/__init__.py]: s/confiruations/configurations

File: cylc/sphinx_ext/rtd_theme_addons/__init__.py [hunk 5]
```diff
@@ -6,6 +6,12 @@
 
    These modifications only work in the `Read The Docs Theme`_
 
+.. note::
+
+   extension is automatically loaded when added to extensions, no directives or
+   confiruations required.
```
Comment (kinow) [inline cylc/sphinx_ext/rtd_theme_addons/__init__.py]: s/confiruations/configurations

File: cylc/sphinx_ext/cylc_lang/__init__.py [hunk 6]
```diff
@@ -34,37 +39,142 @@
...
+
+           a setting called ``pub``
+
+           .. cylc:value:: integer
+
+              :deprecated: 1.2.3
+
+              seconds as an integer.
+
+              the newer :cylc:value:`string` is prefered (this is also
```
Comment (kinow) [inline cylc/sphinx_ext/cylc_lang/__init__.py]: s/prefered/preferred

File: cylc/sphinx_ext/cylc_lang/domains.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+class CylcXRefRole(XRefRole):
+    """Handle references to Rose objects.
+
+    This should be minimal."""
+
+    def process_link(self, env, refnode, has_explicit_title, title, target):
+        # copy ref_context to the refnode so that we can access it in
+        # resolve_xref. Note that walking through the node tree to extract
+        # ref_context items appears only to work in the HTML buider.
```
Comment (kinow) [inline cylc/sphinx_ext/cylc_lang/domains.py]: s/buider/builder

**Comment by oliver-sanders** [comment]
> @oliver-sanders is that expected until cylc/cylc-flow#3253 is done? If so will approve this PR.

Hmmm, curious, `singlehtml` works for me, will try and re-produce.

**Comment by kinow** [comment]
> 
> 
> > @oliver-sanders is that expected until [cylc/cylc-flow#3253](https://github.com/cylc/cylc-flow/issues/3253) is done? If so will approve this PR.
> 
> Hmmm, curious, `singlehtml` works for me, will try and re-produce.

I think the `html` target failed first. But from your previous comment it looks like it worked for you too. I wonder if I am missing something else?

**Comment by oliver-sanders** [comment]
I can't replicate this ATM, here's my attempt to reproduce:

```console
$ conda create -y -n cse
$ conda activate cse
$ conda install -y -c conda-forge python==3.7
$ cd ~/cylc-flow/; pip install -e .[all]
$ cd ~/cylc-sphinx-extensions; pip install -e .[all]
$ conda install graphviz  # one thing we cant package with pip
$ make clean html
```

**Comment by oliver-sanders** [comment]
Ooh, actually try this

```console
$ rm _templates/mymodule.rst
$ make clean html
```

**Comment by kinow** [comment]
> Ooh, actually try this
> 
> ```
> $ rm _templates/mymodule.rst
> $ make clean html
> ```

I was installing `graphviz` when your comment arrived. I had the same error that I had with pip. Then tried your commands above (after `conda install graphviz`). Now got:

```bash
(cse) kinow@kinow-VirtualBox:~/Development/python/workspace/cylc-sphinx-extensions$ rm _templates/mymodule.rst
(cse) kinow@kinow-VirtualBox:~/Development/python/workspace/cylc-sphinx-extensions$ make clean html
rm -r extensions/ || true
Removing everything under '_build'...
# build documentation
Running Sphinx v2.4.1
making output directory... done
[autosummary] generating autosummary for: README.rst, index.rst, venv/lib/python3.7/site-packages/Jinja2-2.11.1.dist-info/LICENSE.rst, venv/lib/python3.7/site-packages/Rx-1.6.1.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/alabaster-0.7.12.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/ansimarkup-1.4.0.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/certifi-2019.11.28.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/chardet-3.0.4.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/codecov-2.0.15.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/hieroglyph-1.0.0.dist-info/DESCRIPTION.rst, ..., venv/lib/python3.7/site-packages/imagesize-1.2.0.dist-info/LICENSE.rst, venv/lib/python3.7/site-packages/importlib_metadata/docs/changelog.rst, venv/lib/python3.7/site-packages/importlib_metadata/docs/index.rst, venv/lib/python3.7/site-packages/importlib_metadata/docs/using.rst, venv/lib/python3.7/site-packages/pycodestyle-2.5.0.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/pytest_cov-2.8.1.dist-info/AUTHORS.rst, venv/lib/python3.7/site-packages/pytz-2019.3.dist-info/DESCRIPTION.rst, venv/lib/python3.7/site-packages/sphinx/ext/autosummary/templates/autosummary/base.rst, venv/lib/python3.7/site-packages/sphinx/ext/autosummary/templates/autosummary/class.rst, venv/lib/python3.7/site-packages/sphinx/ext/autosummary/templates/autosummary/module.rst

Exception occurred:
  File "/home/kinow/anaconda3/envs/cse/lib/python3.7/site-packages/sphinx/jinja2glue.py", line 213, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: mymodule.rst
The full traceback has been saved in /tmp/sphinx-err-ub5fjd5s.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:23: recipe for target 'html' failed
make: *** [html] Error 2

```

**Comment by kinow** [comment]
Checked out the `_templates/mymodule.rst` file again, same error.

```bash
reading sources... [100%] venv/lib/python3.7/site-packages/sphinx/ext/autosummary/templates/autosummary/module                                                                                                     

Warning, treated as error:
/home/kinow/Development/python/workspace/cylc-sphinx-extensions/venv/lib/python3.7/site-packages/importlib_metadata/docs/index.rst:24:toctree contains reference to nonexisting document 'venv/lib/python3.7/site-packages/importlib_metadata/docs/changelog (links)'
Makefile:23: recipe for target 'html' failed
make: *** [html] Error 2
```

**Comment by oliver-sanders** [comment]
(sorry, you need to apply the diff rather than remove the file)

**Comment by kinow** [comment]
>(sorry, you need to apply the diff rather than remove the file, corrected comment above)

Diff applied

image

Still same error (tried with `deactivate; conda activate cse`, and later `conda deactivate; source venv/bin/activate`, both Conda & pip failing).

>Is your virtualenv pulling in Sphinx or Docutils from the base installation?

**venv**

image

**conda**

image

It could be something else in my environment. Maybe another reviewer could try it later and see if that works. I can also try from home later today to check if I get the same error or not?

**Comment by oliver-sanders** [comment]
Those versions match what I'm testing with and the paths look file.

I have no idea what's going wrong for you. The tests are passing on Travis, for info all Travis does to install is this:

```
    - pip install git+https://github.com/cylc/cylc-flow/  # install latest cylc-flow
    - pip install -e .[all]
    - npm install -g eslint
    - sudo apt-get install -y graphviz
```

The only thing left is to examine the full traceback stack, you'll find this in the sphinx output:

```
The full traceback has been saved in /tmp/sphinx-err-ub5fjd5s.log
```

Perhaps post to Riot as it'll be quite long.

**Comment by kinow** [comment]
> The only thing left is to examine the full traceback stack, you'll find this in the sphinx output:
> 
> ```
> The full traceback has been saved in /tmp/sphinx-err-ub5fjd5s.log
> ```
> 
> Perhaps post to Riot as it'll be quite long.

That traceback was from when I did a `rm _templates/mymodule.rst`.

> Those versions match what I'm testing with and the paths look file.
> 
> I have no idea what's going wrong for you. The tests are passing on Travis, for info all Travis does to install is this:
> 
> ```
>     - pip install git+https://github.com/cylc/cylc-flow/  # install latest cylc-flow
>     - pip install -e .[all]
>     - npm install -g eslint
>     - sudo apt-get install -y graphviz

It must be something I'm doing wrong. Just tried at home, and got the same error. Will try to copy the Travis installation steps and see if that works for me. Thanks!

**Comment by kinow** [comment]
@oliver-sanders found it! And the cause was my `venv` folder. Even with the Conda environment, that `venv` directory was still being used — apparently.

I decided to search for that error, and got few uninteresting results. And one in Japanese (normally I see those in Japanese or Chinese and ignore). But this one was spot on, even the solution/explanation was good.

Translated: https://translate.google.com/translate?sl=ja&tl=en&u=http%3A%2F%2Fcointoss.hatenablog.com%2Fentry%2F2013%2F03%2F31%2Fsphinx-including-virtualenv

>Cause
>
>The cause was that the rst file included in sphinx installed under .env was also targeted for build.
>
>If sphinx does not separate the source directory, the current directory is specified as the source directory * 1
>In that case, .env is also specified as the source folder, and an error occurs when trying to build a .rst file that exists deep in .env. 

And the solution worked for me:

>solution
>
>Exclude .env from build targets.
>
>Add .env to exclude_patterns in conf.py. Since it is added as a character string, enclose it in ''. 

But I had to include the following:

```diff
diff --git a/conf.py b/conf.py
index 9d04f02..662df75 100644
--- a/conf.py
+++ b/conf.py
@@ -59,3 +59,8 @@ autosummary_imported_members = True
 intersphinx_mapping = {
     'sphinx': ('http://www.sphinx-doc.org/', None)
 }
+
+# List of patterns, relative to source directory, that match files and
+# directories to ignore when looking for source files.
+exclude_patterns = ['_build', 'venv']
+
```

Approving the PR. But as others could use `pip`, maybe we should add this exclude patterns just to play safe?

Thanks for the help troubleshooting it!

**Comment by oliver-sanders** [comment]
Aah, creating a virtualenv in the root directory, nice detective work, have applied your fix.

File: cylc/sphinx_ext/cylc_lang/__init__.py (hunk 6 above)
Comment (wxtim) [reply cylc/sphinx_ext/cylc_lang/__init__.py]:
```suggestion
              the newer :cylc:value:`string` is preffered (this is also
```

File: cylc/sphinx_ext/diff_selection/__init__.py (hunk 1 above)
Comment (wxtim) [reply cylc/sphinx_ext/diff_selection/__init__.py]:
```suggestion
   configurations required.
```

File: cylc/sphinx_ext/grid_table/__init__.py (hunk 2 above)
Comment (wxtim) [reply cylc/sphinx_ext/grid_table/__init__.py]:
```suggestion
   configurations required.
```

File: cylc/sphinx_ext/hieroglyph_addons/__init__.py (hunk 3 above)
Comment (wxtim) [reply cylc/sphinx_ext/hieroglyph_addons/__init__.py]:
```suggestion
   configurations required.
```

File: cylc/sphinx_ext/hieroglyph_patch/__init__.py (hunk 4 above)
Comment (wxtim) [reply cylc/sphinx_ext/hieroglyph_patch/__init__.py]:
```suggestion
   configurations required.
```

File: cylc/sphinx_ext/rtd_theme_addons/__init__.py (hunk 5 above)
Comment (wxtim) [reply cylc/sphinx_ext/rtd_theme_addons/__init__.py]:
```suggestion
   configurations required.
```

**Comment by oliver-sanders** [comment]
> How long should the "singlehtml" output be? It looked a bit short to me:

Never tried singlehtml, the "stub" pages don't seem to get included, no idea, shouldn't matter though.

**Comment by kinow** [comment]
`.travis.yml` conflicts

**Comment by kinow** [comment]
Two approvals, and all checks passing. LGTM

**Pull request merged**